// gatsby-source-sanity/lib-es5/gatsby-node.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setFieldsOnGraphQLNodeType = exports.onPreExtractQueries = exports.sourceNodes = exports.createSchemaCustomization = exports.createResolvers = exports.onPreBootstrap = exports.onPreInit = void 0;
var fs_extra_1 = require("fs-extra");
var operators_1 = require("rxjs/operators");
var package_json_1 = __importDefault(require("gatsby/package.json"));
var client_1 = __importDefault(require("@sanity/client"));
var normalize_1 = require("./util/normalize");
var cache_1 = require("./util/cache");
var handleWebhookEvent_1 = require("./util/handleWebhookEvent");
var remoteGraphQLSchema_1 = require("./util/remoteGraphQLSchema");
var errors_1 = require("./util/errors");
var extendImageNode_1 = require("./images/extendImageNode");
var rewriteGraphQLSchema_1 = require("./util/rewriteGraphQLSchema");
var getGraphQLResolverMap_1 = require("./util/getGraphQLResolverMap");
var documentIds_1 = require("./util/documentIds");
var getAllDocuments_1 = require("./util/getAllDocuments");
var path_1 = __importDefault(require("path"));
var oneline_1 = __importDefault(require("oneline"));
var lodash_1 = require("lodash");
var debug_1 = __importDefault(require("./debug"));
var defaultConfig = {
    version: '1',
    overlayDrafts: false,
    graphqlTag: 'default',
};
var stateCache = {};
var onPreInit = function (_a) {
    var reporter = _a.reporter;
    return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_b) {
            // Old versions of Gatsby does not have this method
            if (reporter.setErrorMap) {
                reporter.setErrorMap(errors_1.ERROR_MAP);
            }
            return [2 /*return*/];
        });
    });
};
exports.onPreInit = onPreInit;
var onPreBootstrap = function (args, pluginOptions) { return __awaiter(void 0, void 0, void 0, function () {
    var config, reporter, unsupportedVersionMessage, client, api, graphqlSdl, graphqlSdlKey, typeMap, typeMapKey, err_1, prefix;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                config = __assign(__assign({}, defaultConfig), pluginOptions);
                reporter = args.reporter;
                if (Number(package_json_1.default.version.split('.')[0]) < 3) {
                    unsupportedVersionMessage = oneline_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    You are using a version of Gatsby not supported by gatsby-source-sanity.\n    Upgrade gatsby to >= 3.0.0 to continue."], ["\n    You are using a version of Gatsby not supported by gatsby-source-sanity.\n    Upgrade gatsby to >= 3.0.0 to continue."])));
                    reporter.panic({
                        id: errors_1.prefixId(errors_1.ERROR_CODES.UnsupportedGatsbyVersion),
                        context: { sourceMessage: unsupportedVersionMessage },
                    });
                    return [2 /*return*/];
                }
                // Actually throws in validation function, but helps typescript perform type narrowing
                if (!validateConfig(config, reporter)) {
                    throw new Error('Invalid config');
                }
                _a.label = 1;
            case 1:
                _a.trys.push([1, 4, , 5]);
                reporter.info('[sanity] Fetching remote GraphQL schema');
                client = getClient(config);
                return [4 /*yield*/, remoteGraphQLSchema_1.getRemoteGraphQLSchema(client, config)];
            case 2:
                api = _a.sent();
                reporter.info('[sanity] Transforming to Gatsby-compatible GraphQL SDL');
                return [4 /*yield*/, rewriteGraphQLSchema_1.rewriteGraphQLSchema(api, { config: config, reporter: reporter })];
            case 3:
                graphqlSdl = _a.sent();
                graphqlSdlKey = cache_1.getCacheKey(config, cache_1.CACHE_KEYS.GRAPHQL_SDL);
                stateCache[graphqlSdlKey] = graphqlSdl;
                reporter.info('[sanity] Stitching GraphQL schemas from SDL');
                typeMap = remoteGraphQLSchema_1.getTypeMapFromGraphQLSchema(api);
                typeMapKey = cache_1.getCacheKey(config, cache_1.CACHE_KEYS.TYPE_MAP);
                stateCache[typeMapKey] = typeMap;
                return [3 /*break*/, 5];
            case 4:
                err_1 = _a.sent();
                if (err_1.isWarning) {
                    err_1.message.split('\n').forEach(function (line) { return reporter.warn(line); });
                    return [2 /*return*/];
                }
                if (typeof err_1.code === 'string' && errors_1.SANITY_ERROR_CODE_MAP[err_1.code]) {
                    reporter.panic({
                        id: documentIds_1.prefixId(errors_1.SANITY_ERROR_CODE_MAP[err_1.code]),
                        context: { sourceMessage: "[sanity] " + errors_1.SANITY_ERROR_CODE_MESSAGES[err_1.code] },
                    });
                }
                prefix = typeof err_1.code === 'string' ? "[" + err_1.code + "] " : '';
                reporter.panic({
                    id: documentIds_1.prefixId(errors_1.ERROR_CODES.SchemaFetchError),
                    context: { sourceMessage: "" + prefix + err_1.message },
                });
                return [3 /*break*/, 5];
            case 5: return [2 /*return*/];
        }
    });
}); };
exports.onPreBootstrap = onPreBootstrap;
var createResolvers = function (args, pluginOptions) {
    var typeMapKey = cache_1.getCacheKey(pluginOptions, cache_1.CACHE_KEYS.TYPE_MAP);
    var typeMap = (stateCache[typeMapKey] || remoteGraphQLSchema_1.defaultTypeMap);
    args.createResolvers(getGraphQLResolverMap_1.getGraphQLResolverMap(typeMap, pluginOptions, args));
};
exports.createResolvers = createResolvers;
var createSchemaCustomization = function (_a, pluginConfig) {
    var actions = _a.actions;
    var createTypes = actions.createTypes;
    var graphqlSdlKey = cache_1.getCacheKey(pluginConfig, cache_1.CACHE_KEYS.GRAPHQL_SDL);
    var graphqlSdl = stateCache[graphqlSdlKey];
    createTypes(graphqlSdl);
};
exports.createSchemaCustomization = createSchemaCustomization;
var sourceNodes = function (args, pluginConfig) { return __awaiter(void 0, void 0, void 0, function () {
    // sync a single document from the local cache of known documents with gatsby
    function syncWithGatsby(id) {
        var publishedId = documentIds_1.unprefixId(id);
        var draftId = documentIds_1.prefixId(id);
        var published = documents.get(publishedId);
        var draft = documents.get(draftId);
        var doc = draft || published;
        if (doc) {
            var type = normalize_1.getTypeName(doc._type);
            if (!typeMap.objects[type]) {
                reporter.warn("[sanity] Document \"" + doc._id + "\" has type " + doc._type + " (" + type + "), which is not declared in the GraphQL schema. Make sure you run \"graphql deploy\". Skipping document.");
                return;
            }
        }
        if (id === draftId && !overlayDrafts) {
            // do nothing, we're not overlaying drafts
            debug_1.default('overlayDrafts is not enabled, so skipping createNode for draft');
            return;
        }
        if (id === publishedId) {
            if (draft && overlayDrafts) {
                // we have a draft, and overlayDrafts is enabled, so skip to the draft document instead
                debug_1.default('skipping createNode of %s since there is a draft and overlayDrafts is enabled', publishedId);
                return;
            }
            if (gatsbyNodes.has(publishedId)) {
                // sync existing gatsby node with document from updated cache
                if (published) {
                    debug_1.default('updating gatsby node for %s', publishedId);
                    var node = normalize_1.toGatsbyNode(published, processingOptions);
                    gatsbyNodes.set(publishedId, node);
                    createNode(node);
                }
                else {
                    // the published document has been removed (note - we either have no draft or overlayDrafts is not enabled so merely removing is ok here)
                    debug_1.default('deleting gatsby node for %s since there is no draft and overlayDrafts is not enabled', publishedId);
                    deleteNode(gatsbyNodes.get(publishedId));
                    gatsbyNodes.delete(publishedId);
                }
            }
            else if (published) {
                // when we don't have a gatsby node for the published document
                debug_1.default('creating gatsby node for %s', publishedId);
                var node = normalize_1.toGatsbyNode(published, processingOptions);
                gatsbyNodes.set(publishedId, node);
                createNode(node);
            }
        }
        if (id === draftId && overlayDrafts) {
            // we're syncing a draft version and overlayDrafts is enabled
            if (gatsbyNodes.has(publishedId) && !draft && !published) {
                // have stale gatsby node for a published document that has neither a draft or a published (e.g. it's been deleted)
                debug_1.default('deleting gatsby node for %s since there is neither a draft nor a published version of it any more', publishedId);
                deleteNode(gatsbyNodes.get(publishedId));
                gatsbyNodes.delete(publishedId);
                return;
            }
            debug_1.default('Replacing gatsby node for %s using the %s document', publishedId, draft ? 'draft' : 'published');
            // pick the draft if we can, otherwise pick the published
            var node = normalize_1.toGatsbyNode((draft || published), processingOptions);
            gatsbyNodes.set(publishedId, node);
            createNode(node);
        }
    }
    function syncAllWithGatsby() {
        var e_1, _a;
        try {
            for (var _b = __values(documents.keys()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var id = _c.value;
                syncWithGatsby(id);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    function syncIdsWithGatsby(ids) {
        var e_2, _a;
        try {
            for (var ids_1 = __values(ids), ids_1_1 = ids_1.next(); !ids_1_1.done; ids_1_1 = ids_1.next()) {
                var id = ids_1_1.value;
                syncWithGatsby(id);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (ids_1_1 && !ids_1_1.done && (_a = ids_1.return)) _a.call(ids_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    }
    var config, dataset, overlayDrafts, watchMode, actions, createNodeId, createContentDigest, reporter, webhookBody, createNode, deleteNode, createParentChildLink, typeMapKey, typeMap, client, url, processingOptions, _a, documents, gatsbyNodes;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                config = __assign(__assign({}, defaultConfig), pluginConfig);
                dataset = config.dataset, overlayDrafts = config.overlayDrafts, watchMode = config.watchMode;
                actions = args.actions, createNodeId = args.createNodeId, createContentDigest = args.createContentDigest, reporter = args.reporter, webhookBody = args.webhookBody;
                createNode = actions.createNode, deleteNode = actions.deleteNode, createParentChildLink = actions.createParentChildLink;
                typeMapKey = cache_1.getCacheKey(pluginConfig, cache_1.CACHE_KEYS.TYPE_MAP);
                typeMap = (stateCache[typeMapKey] || remoteGraphQLSchema_1.defaultTypeMap);
                client = getClient(config);
                url = client.getUrl("/data/export/" + dataset);
                processingOptions = {
                    typeMap: typeMap,
                    createNodeId: createNodeId,
                    createNode: createNode,
                    createContentDigest: createContentDigest,
                    createParentChildLink: createParentChildLink,
                    overlayDrafts: overlayDrafts,
                };
                _a = webhookBody &&
                    webhookBody.ids;
                if (!_a) return [3 /*break*/, 2];
                return [4 /*yield*/, handleWebhookEvent_1.handleWebhookEvent(args, { client: client, processingOptions: processingOptions })];
            case 1:
                _a = (_b.sent());
                _b.label = 2;
            case 2:
                if (_a) {
                    // If the payload was handled by the webhook handler, fall back.
                    // Otherwise, this may not be a Sanity webhook, but we should
                    // still attempt to refresh our data
                    return [2 /*return*/];
                }
                reporter.info('[sanity] Fetching export stream for dataset');
                return [4 /*yield*/, downloadDocuments(url, config.token, { includeDrafts: overlayDrafts })];
            case 3:
                documents = _b.sent();
                gatsbyNodes = new Map();
                if (watchMode) {
                    // Note: since we don't setup the listener before *after* all documents has been fetched here we will miss any events that
                    // happened in the time window between the documents was fetched and the listener connected. If this happens, the
                    // preview will show an outdated version of the document.
                    reporter.info('[sanity] Watch mode enabled, starting a listener');
                    client
                        .listen('*[!(_id in path("_.**"))]')
                        .pipe(operators_1.filter(function (event) { return overlayDrafts || !event.documentId.startsWith('drafts.'); }), operators_1.tap(function (event) {
                        if (event.result) {
                            documents.set(event.documentId, event.result);
                        }
                        else {
                            documents.delete(event.documentId);
                        }
                    }), operators_1.map(function (event) { return event.documentId; }), operators_1.bufferTime(100), operators_1.map(function (ids) { return lodash_1.uniq(ids); }), operators_1.filter(function (ids) { return ids.length > 0; }), operators_1.tap(function (updateIds) {
                        return debug_1.default('The following documents updated and will be synced with gatsby: ', updateIds);
                    }), operators_1.tap(function (updatedIds) { return syncIdsWithGatsby(updatedIds); }))
                        .subscribe();
                }
                // do the initial sync from sanity documents to gatsby nodes
                syncAllWithGatsby();
                reporter.info("[sanity] Done! Exported " + documents.size + " documents.");
                return [2 /*return*/];
        }
    });
}); };
exports.sourceNodes = sourceNodes;
var onPreExtractQueries = function (context, pluginConfig) { return __awaiter(void 0, void 0, void 0, function () {
    var getNodes, store, typeMapKey, typeMap, shouldAddFragments, program;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                getNodes = context.getNodes, store = context.store;
                typeMapKey = cache_1.getCacheKey(pluginConfig, cache_1.CACHE_KEYS.TYPE_MAP);
                typeMap = (stateCache[typeMapKey] || remoteGraphQLSchema_1.defaultTypeMap);
                shouldAddFragments = typeof typeMap.objects.SanityImageAsset !== 'undefined';
                if (!shouldAddFragments) {
                    shouldAddFragments = getNodes().some(function (node) {
                        return Boolean(node.internal && node.internal.type === 'SanityImageAsset');
                    });
                }
                if (!shouldAddFragments) return [3 /*break*/, 2];
                program = store.getState().program;
                return [4 /*yield*/, fs_extra_1.copy(path_1.default.join(__dirname, '..', 'fragments', 'imageFragments.js'), program.directory + "/.cache/fragments/sanity-image-fragments.js")];
            case 1:
                _a.sent();
                _a.label = 2;
            case 2: return [2 /*return*/];
        }
    });
}); };
exports.onPreExtractQueries = onPreExtractQueries;
var setFieldsOnGraphQLNodeType = function (context, pluginConfig) { return __awaiter(void 0, void 0, void 0, function () {
    var type, fields;
    return __generator(this, function (_a) {
        type = context.type;
        fields = {};
        if (type.name === 'SanityImageAsset') {
            fields = __assign(__assign({}, fields), extendImageNode_1.extendImageNode(pluginConfig));
        }
        return [2 /*return*/, fields];
    });
}); };
exports.setFieldsOnGraphQLNodeType = setFieldsOnGraphQLNodeType;
function validateConfig(config, reporter) {
    if (!config.projectId) {
        reporter.panic({
            id: documentIds_1.prefixId(errors_1.ERROR_CODES.MissingProjectId),
            context: { sourceMessage: '[sanity] `projectId` must be specified' },
        });
    }
    if (!config.dataset) {
        reporter.panic({
            id: documentIds_1.prefixId(errors_1.ERROR_CODES.MissingDataset),
            context: { sourceMessage: '[sanity] `dataset` must be specified' },
        });
    }
    if (config.overlayDrafts && !config.token) {
        reporter.warn('[sanity] `overlayDrafts` is set to `true`, but no token is given');
    }
    var inDevelopMode = process.env.gatsby_executing_command === 'develop';
    if (config.watchMode && !inDevelopMode) {
        reporter.warn('[sanity] Using `watchMode` when not in develop mode might prevent your build from completing');
    }
    return true;
}
function downloadDocuments(url, token, options) {
    if (options === void 0) { options = {}; }
    return getAllDocuments_1.getAllDocuments(url, token, options).then(function (stream) {
        return new Promise(function (resolve, reject) {
            var documents = new Map();
            stream.on('data', function (doc) {
                documents.set(doc._id, doc);
            });
            stream.on('end', function () {
                resolve(documents);
            });
            stream.on('error', function (error) {
                reject(error);
            });
        });
    });
}
function getClient(config) {
    var projectId = config.projectId, dataset = config.dataset, token = config.token;
    return new client_1.default({
        projectId: projectId,
        dataset: dataset,
        token: token,
        useCdn: false,
    });
}
var templateObject_1;
