// @react-three/drei/core/index.js
import 'blob-polyfill';
import '../helpers/base64.js';
import '../assets/cloud.base64.js';
import '@babel/runtime/helpers/esm/extends';
import '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import 'react';
import 'react-three-fiber';
export { B as Billboard, a0 as Box, c as CameraShake, t as Center, a1 as Circle, C as Cloud, a2 as Cone, $ as ContactShadows, d as CubeCamera, Q as CurveModifier, a3 as Cylinder, D as Detailed, e as DeviceOrientationControls, ae as Dodecahedron, a as Effects, E as Environment, af as Extrude, F as FlyControls, G as GizmoHelper, k as GizmoViewcube, l as GizmoViewport, ac as Icosahedron, ag as Lathe, L as Line, M as MapControls, U as MeshDistortMaterial, V as MeshWobbleMaterial, ad as Octahedron, f as OrbitControls, O as OrthographicCamera, ah as Parametric, b as PerspectiveCamera, a5 as Plane, j as PointerLockControls, ab as Polyhedron, P as PositionalAudio, v as Preload, R as Reflector, aa as Ring, ai as RoundedBox, aj as ScreenQuad, S as Shadow, Y as Sky, a4 as Sphere, _ as Stars, s as Stats, a9 as Tetrahedron, T as Text, a7 as Torus, a8 as TorusKnot, g as TrackballControls, h as TransformControls, a6 as Tube, X as calcPosFromAngles, i as isWebGL2Available, r as meshBounds, W as shaderMaterial, Z as softShadows, H as useAnimations, w as useAspect, x as useCamera, A as useContextBridge, m as useCubeTexture, y as useDetectGPU, N as useEdgeSplit, I as useFBO, n as useFBX, o as useGLTF, u as useGizmoContext, z as useHelper, ak as useMatcapTexture, al as useNormalTexture, p as useProgress, J as useSimplification, K as useTessellation, q as useTexture } from '../useNormalTexture-5b7cd1b1.js';
import 'react-merge-refs';
import 'three';
import 'three/examples/jsm/lines/LineGeometry';
import 'three/examples/jsm/lines/LineMaterial';
import 'three/examples/jsm/lines/Line2';
import 'troika-three-text';
import 'three/examples/jsm/loaders/RGBELoader';
import 'use-asset';
import '../helpers/environment-assets.js';
import 'three/examples/jsm/postprocessing/EffectComposer';
import 'three/examples/jsm/postprocessing/RenderPass';
import 'three/examples/jsm/postprocessing/ShaderPass';
import 'three/examples/jsm/shaders/GammaCorrectionShader';
import 'three/examples/jsm/math/SimplexNoise';
import 'three/examples/jsm/controls/DeviceOrientationControls';
import 'three/examples/jsm/controls/FlyControls';
import 'three/examples/jsm/controls/OrbitControls';
import 'three/examples/jsm/controls/TrackballControls';
import 'three/examples/jsm/controls/TransformControls';
import 'lodash.pick';
import 'lodash.omit';
import 'three/examples/jsm/controls/PointerLockControls';
import 'three/examples/jsm/loaders/FBXLoader';
import 'three/examples/jsm/loaders/GLTFLoader';
import 'three/examples/jsm/loaders/DRACOLoader';
import 'three/examples/jsm/libs/meshopt_decoder.module.js';
import 'zustand';
import '@babel/runtime/helpers/esm/defineProperty';
import '../MeshReflectorMaterial-fdc52c9a.js';
import 'three/examples/js/libs/stats.min';
import 'detect-gpu';
import 'three/examples/jsm/modifiers/SimplifyModifier';
import 'three/examples/jsm/modifiers/TessellateModifier';
import 'three/examples/jsm/modifiers/EdgeSplitModifier';
import 'three/examples/jsm/modifiers/CurveModifier';
import 'three/examples/jsm/objects/Sky';
import 'three/examples/jsm/shaders/HorizontalBlurShader';
import 'three/examples/jsm/shaders/VerticalBlurShader';
import '../helpers/matcap-assets.js';
import '../helpers/normal-assets.js';
